File: server-go/internal/ai/ai_test.go
package ai

import (
	"testing"

	"github.com/flowforge/flowforge/internal/models"
)

// Go mirror of AI-01/AI-02 (deterministic path only). Feature F-AI.
func TestDeterministicFallback(t *testing.T) {
	cfg := models.AIConfig{Provider: "openai", APIKey: "", BaseURL: "https://api.openai.com/v1", Model: "gpt-4o-mini"}
	res := GenerateDraft("When a vendor invoice over $10K arrives, extract line items, validate against the vendor master, route to the manager for approval, then post to the ERP.", cfg)

	if res.Source != "fallback" {
		t.Errorf("source = %q, want fallback", res.Source)
	}
	d := res.Draft
	if len(d.Steps) < 2 {
		t.Fatalf("too few steps: %d", len(d.Steps))
	}
	if d.Steps[0].Type != "trigger" {
		t.Errorf("first step must be trigger, got %q", d.Steps[0].Type)
	}
	triggers := 0
	for _, s := range d.Steps {
		if s.Type == "trigger" {
			triggers++
		}
	}
	if triggers != 1 {
		t.Errorf("want exactly 1 trigger, got %d", triggers)
	}
	hasCond, hasAppr, hasPost := false, false, false
	for _, s := range d.Steps {
		switch s.Type {
		case "condition":
			hasCond = true
		case "human.approval":
			hasAppr = true
		case "integration.post":
			hasPost = true
		}
	}
	if !hasCond || !hasAppr || !hasPost {
		t.Errorf("expected condition + approval + post: %v %v %v", hasCond, hasAppr, hasPost)
	}
	for _, s := range d.Steps {
		if s.Confidence < 50 || s.Confidence > 99 {
			t.Errorf("confidence out of range: %d", s.Confidence)
		}
		if s.Assumptions == nil {
			t.Errorf("step %q assumptions must not be nil (JSON contract: [])", s.ID)
		}
		if s.Params == nil {
			t.Errorf("step %q params must not be nil (JSON contract: {})", s.ID)
		}
	}
	if d.OverallConfidence <= 0 {
		t.Errorf("overall confidence should be positive")
	}
}

File: server/tests/ai.test.ts
import { describe, it, expect } from 'vitest';
import { generateDraft } from '../src/ai.js';
import type { AIConfig } from '../src/settings.js';

// Scenario IDs: AI-01, AI-02 (see docs/test-strategy.md). Feature F-AI.
// Deterministic path only (no model configured) — LLM live tests are separate.

const inactive: AIConfig = { provider: 'openai', apiKey: '', baseURL: 'https://api.openai.com/v1', model: 'gpt-4o-mini' };

const KNOWN = new Set([
  'trigger', 'ai.extract', 'ai.classify', 'mdm.lookup', 'mdm.validate', 'condition',
  'human.approval', 'notify', 'integration.post', 'integration.http', 'script', 'wait',
]);

const PROMPT = 'When a vendor invoice over $10K arrives, extract line items, validate against the vendor master, route to the manager for approval, then post to the ERP.';

describe('ai authoring (AI)', () => {
  it('AI-01 falls back to a deterministic, well-formed draft when no model is configured', async () => {
    const res = await generateDraft(PROMPT, inactive);
    expect(res.source).toBe('fallback');
    expect(res.draft.steps.length).toBeGreaterThan(1);
    // exactly one trigger, and it is first
    const triggers = res.draft.steps.filter((s) => s.type === 'trigger');
    expect(triggers).toHaveLength(1);
    expect(res.draft.steps[0].type).toBe('trigger');
    // every step type is a known control
    for (const s of res.draft.steps) expect(KNOWN.has(s.type)).toBe(true);
    // confidence in [50,99]
    for (const s of res.draft.steps) {
      expect(s.confidence).toBeGreaterThanOrEqual(50);
      expect(s.confidence).toBeLessThanOrEqual(99);
    }
    expect(res.draft.overallConfidence).toBeGreaterThan(0);
  });

  it('AI-02 infers condition + human approval from the prompt', async () => {
    const res = await generateDraft(PROMPT, inactive);
    expect(res.draft.steps.some((s) => s.type === 'condition')).toBe(true);
    expect(res.draft.steps.some((s) => s.type === 'human.approval')).toBe(true);
    expect(res.draft.steps.some((s) => s.type === 'integration.post')).toBe(true);
  });
});
